diff --git a/pkgs/development/python-modules/kornia/default.nix b/pkgs/development/python-modules/kornia/default.nix
new file mode 100644
index 000000000000..528628f75d87
--- /dev/null
+++ b/pkgs/development/python-modules/kornia/default.nix
@@ -0,0 +1,53 @@
+{ buildPythonPackage
+, fetchFromGitHub
+, lib
+
+# propagated build inputs
+, packaging
+, pytorch
+
+# check inputs
+, pytestCheckHook
+, opencv3
+, scipy
+}:
+
+buildPythonPackage rec {
+  pname = "kornia";
+  version = "0.6.10";
+
+  src = fetchFromGitHub {
+    owner = "kornia";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-4TE9GfcRMSpI18IEgAeuKjTroKGknc2dVlmXOdyiKuE";
+  };
+
+  postPatch = ''
+    substituteInPlace setup.py \
+      --replace "'pytest-runner'" ""
+  '';
+
+  checkInputs = [
+    pytestCheckHook
+    opencv3
+    scipy
+  ];
+
+  propagatedBuildInputs = [
+    packaging
+    pytorch
+  ];
+
+  # tests are extremely slow
+  doCheck = false;
+
+  pythonImportsCheck = ["kornia"];
+
+  meta = with lib; {
+    description = "Differentiable computer vision library for PyTorch";
+    homepage = "https://kornia.readthedocs.io/";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ rehno-lindeque ];
+  };
+}
diff --git a/pkgs/top-level/python-packages.nix b/pkgs/top-level/python-packages.nix
index 21e75c132962..ecfcda61ab24 100644
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -5252,6 +5252,8 @@ self: super: with self; {
 
   korean-lunar-calendar = callPackage ../development/python-modules/korean-lunar-calendar { };
 
+  kornia = callPackage ../development/python-modules/kornia { };
+
   krakenex = callPackage ../development/python-modules/krakenex { };
 
   kubernetes = callPackage ../development/python-modules/kubernetes { };
